<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · EvoDynamics.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">EvoDynamics.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li><a class="tocitem" href="../tutorial/">Tutorial</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Parameters"><span>Parameters</span></a></li><li><a class="tocitem" href="#Simulation-outline"><span>Simulation outline</span></a></li><li><a class="tocitem" href="#Data-collection"><span>Data collection</span></a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../example1/">Simple Wright-Fisher</a></li><li><a class="tocitem" href="../example2/">Predator prey</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/kavir1698/EvoDynamics.jl/blob/master/docs/src/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><h2 id="Parameters"><a class="docs-heading-anchor" href="#Parameters">Parameters</a><a id="Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Parameters" title="Permalink"></a></h2><p>Parameters of a model should be put in a YAML file with the structure below. Note that spaces and indentations are meaningful in YAML. Indentations should be spaces not tabs.</p><p>See <a href="../example1/#Simple-Wright-Fisher">Simple Wright-Fisher</a> and <a href="../example2/#Predator-prey">Predator prey</a> for complete parameter file examples.</p><pre><code class="language-yml hljs">species:
  1:
    name: a
    parameter 1: ...
    parameter 2: ...
    ...
  2:
    name: b
    parameter 1: ... 
    parameter 2: ... 
    ...
model:
  model parameter 1: ...
  model parameter 2: ...
  ...</code></pre><p>The file has two main levels: <code>species</code> and <code>model</code>. <code>species</code> stores species specific parameters as many different species as you want, starting with <code>id</code> 1. </p><p>Since we cannot write a matrix in a YAML file, any parameter that is a matrix should be converted to a vector. In Julia, you can do this by <code>vec(yourmatrix)</code>.</p><h3 id="Species-specific-parameters"><a class="docs-heading-anchor" href="#Species-specific-parameters">Species specific parameters</a><a id="Species-specific-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Species-specific-parameters" title="Permalink"></a></h3><p>Each species should have the following parameters. The order that you write these parameters does not matter.</p><ul><li><strong>number of genes</strong>: An <em>integer</em> for number of genes that the species has.</li><li><strong>ploidy</strong>: Either 1 for haploid or 2 for diploid genomes.</li><li><strong>number of phenotypes</strong>: An <em>integer</em> for the number of phenotypes that the species has.</li><li><strong>abiotic phenotypes</strong>: An <em>array of integers</em> (e.g. &quot;[1,2]&quot;) specifying abiotic phenotypes among all phenotypes. Abiotic phenotypes determine how the species interacts with the environment.</li><li><strong>biotic phenotypes</strong>: An <em>array of integers</em> (e.g. &quot;[3]&quot;) specifying biotic phenotypes among all phenotypes. Biotic phenotypes determine how the species interacts with other individuals from the same or different species.</li><li><strong>migration phenotype</strong>: An <em>integer</em> specifying the phenotype that determines migration trait. If the species does not migrate, put 0.</li><li><strong>vision radius</strong>: A <em>number</em> determining the radius of neighboring sites that the agent can see before migration.</li><li><strong>check fraction</strong>: A <em>number</em> between 0 and 1 showing the fraction of the visible sites to the agent that it can check and decide whether to migrate to.</li><li><strong>epistasis matrix</strong>: An epistasis matrix is of size <span>$l \times l$</span>, where <em>l</em> is the product of <em>number of genes</em> and <em>ploidy</em>. Epistasis matrix specifies the direction (positive or negative) and size of effect of one locus on other loci. For example, if at row 1 and column 2 is a value 0.2, it means that locus 1 affects locus 2 by increasing the effect of locus 2 (because its positive) with 20% of the effect of locus 1.</li><li><strong>pleiotropy matrix</strong>: A <em>binary matrix</em> (0s and 1s) with size <em>number of phenotypes</em> times <em>l</em>. The pleiotropy matrix specifies the phenotypes that each locus affects.</li><li><strong>expression array</strong>: A <em>vector</em> of size <em>l</em> that represent the expression amount of each locus determining its effect size.</li><li><strong>growth rate</strong>: Mean of a Poisson distribution for number of offsprings per reproduction. This number is the maximum mean when fitness of a haploid individual is 1, or the distance between the biotic phenotypes of two diploid individuals is 0.</li><li><strong>selection coefficient</strong>: A number between 0 and 1 that determines the importance of fitness. 0 would be a model without selection.</li><li><strong>mutation probabilities</strong>: A <em>vector of three numbers</em> each of which specifies the probability for a different type of mutations: mutation probability of the <em>expression array</em>, <em>pleiotropy matrix</em>, and <em>epistasis matrix</em>, respectively. </li><li><strong>mutation magnitudes</strong>: A <em>vector of numbers</em> with the same size as <em>mutation probabilities</em> that determines the magnitude of mutation for each of the three categories. Specifically, the numbers are the variances of normal distributions with mean 0 for expression array and epistasis matrices, and probability of changing a 0 and 1 in in the pleiotropy matrix.</li><li><strong>N</strong>: A <em>vector of integers</em> for the initial number of individuals at each site.</li><li><strong>environmental noise</strong>: A number for the variance of a normal distribution with mean 0 that will be added to the phenotypes.</li><li><strong>optimal phenotype values</strong>: One or more vectors of numbers. Each vector should start with a dash and one indentation level. Each vector is the optimal phenotypes for each site for all abiotic traits. There are as many element as number of sites times number of abiotic traits. The first N elements are for the first abiotic trait, where N is the number of sites, and so on. </li><li><strong>optimal phenotypes</strong>: A <em>vector of integers</em> for optimal phenotype indices for each generation, including generation zero.</li><li><strong>age</strong>: An <em>integer</em> for maximum age of individuals of this species.</li><li><strong>recombination</strong>: Mean of a Poisson distributions for number of crossing overs per sexual reproduction.</li><li><strong>initial energy</strong>: A parameter for parental care of infants. Values more than 0 indicate that newly born individuals can survive for a number of times without requiring food from the environment/other species. The consumption rate (i.e. how many generations this initial energy suffices) is determined by the sum of the corresponding rows in &quot;food sources&quot; model parameter.</li></ul><h3 id="Model-parameters"><a class="docs-heading-anchor" href="#Model-parameters">Model parameters</a><a id="Model-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Model-parameters" title="Permalink"></a></h3><ul><li><strong>generations</strong>: An <em>integer</em> for the number of steps the model will run.</li><li><strong>space</strong>: A <em>vector of two integers</em> that determine the size of a grid for space.</li><li><strong>metric</strong>: Either &quot;chebyshev&quot; or &quot;euclidian&quot;. Determines how many neighbors a space site has. &quot;chebyshev&quot; metric means that the r-neighborhood of a position are all positions within the hypercube having side length of 2*floor(r) and being centered in the origin position. &quot;euclidean&quot; metric means that the r-neighborhood of a position are all positions whose cartesian indices have Euclidean distance ≤ r from the cartesian index of the given position.</li><li><strong>periodic</strong>: _Boolean__ (true or false) to determine whether boundaries of the space are connected or not.</li><li><strong>resources</strong>: A <em>vector of integers</em> determining available resources (e.g. vegetation) per site per time step. </li><li><strong>interactions</strong>: A species-species interaction <em>matrix of numbers</em> determining how individuals from different species interact. Each value  is strength of interaction (between 0 and 1). Sign (+/-) is the direction of interaction where positive means similar individuals interact more strongly and negative is dissimilar ones tend to interact more.</li><li><strong>food sources</strong>: A species-species food <em>matrix of numbers</em> determining what each species feeds on (consumption rate). Has priority over interactions. Non-zero diagonal means the food resource is from the environment. It will be read from rows (species order) to columns (species order). </li><li><strong>seed</strong>: Either an <em>integer</em> or <em>Null</em> for random number generator seed.</li></ul><h2 id="Simulation-outline"><a class="docs-heading-anchor" href="#Simulation-outline">Simulation outline</a><a id="Simulation-outline-1"></a><a class="docs-heading-anchor-permalink" href="#Simulation-outline" title="Permalink"></a></h2><p>The simulations are fully agent-based, meaning that agents do not receive any model-level knowledge for what happens to them. The following steps happen in order to agents that are activated in a random order.</p><ol><li>Grow one step older.</li><li>Migrate.</li><li>Eat basic food if the species can.</li><li>Consume energy.</li><li>Interact with other species.<ol><li>If meeting another individual of the same species but with different sex, reproduce.</li></ol></li><li>If haploid, reproduce.</li><li>Survive if not tool old, has energy and by chance given its fitness.</li></ol><h3 id="Mutation"><a class="docs-heading-anchor" href="#Mutation">Mutation</a><a id="Mutation-1"></a><a class="docs-heading-anchor-permalink" href="#Mutation" title="Permalink"></a></h3><p>Mutation can happen at three levels: changing the expression of each gene expressionArrays, changing the pleiotropy matrix pleiotropyMat, and changing the epistasis interactions between genes. The probability that a mutation occurs at each of these levels is controlled by parameter <em>mutation probabilities</em>. And size of mutations when they occur are controlled by parameter mutation magnitudes. The genotype vector <em>y</em> and pleiotropy matrix <em>B</em> of each individual mutates.</p><p>Epistasis matrix <em>A</em> and expression vectors <em>Q</em> mutate by adding their values to random numbers from a normal distribution with mean 0 and standard deviation given in parameter mutMagnitudes.</p><p><em>B</em> mutates by randomly switching 0s and 1s with probability given in parameter mutMagnitudes.</p><h3 id="Fitness-update"><a class="docs-heading-anchor" href="#Fitness-update">Fitness update</a><a id="Fitness-update-1"></a><a class="docs-heading-anchor-permalink" href="#Fitness-update" title="Permalink"></a></h3><p>TODO</p><h3 id="Migration"><a class="docs-heading-anchor" href="#Migration">Migration</a><a id="Migration-1"></a><a class="docs-heading-anchor-permalink" href="#Migration" title="Permalink"></a></h3><p>TODO</p><h3 id="Reproduction"><a class="docs-heading-anchor" href="#Reproduction">Reproduction</a><a id="Reproduction-1"></a><a class="docs-heading-anchor-permalink" href="#Reproduction" title="Permalink"></a></h3><p>TODO</p><h3 id="Survival"><a class="docs-heading-anchor" href="#Survival">Survival</a><a id="Survival-1"></a><a class="docs-heading-anchor-permalink" href="#Survival" title="Permalink"></a></h3><p>TODO</p><h2 id="Data-collection"><a class="docs-heading-anchor" href="#Data-collection">Data collection</a><a id="Data-collection-1"></a><a class="docs-heading-anchor-permalink" href="#Data-collection" title="Permalink"></a></h2><p>The interface to the model is from the <code>runmodel</code> function (see <a href="../tutorial/#Tutorial">Tutorial</a>).</p><p>EvoDynamics.jl uses <a href="https://github.com/JuliaDynamics/Agents.jl">Agents.jl</a> underneath. See <a href="https://juliadynamics.github.io/Agents.jl/dev/">Agents.jl&#39;s documentation</a> for details about writing functions to collect any data during simulations. Here, we explain the specific implementation of the model.</p><p>There are two main objects from which you can collect data: and agent object of type <code>AbstractAgent</code> and a model object of type <code>ABM</code>. Both of these types are defined the <code>Agents.jl</code> package.</p><p>Agent object has the following fields that are individual specific: <code>id</code>, <code>pos</code>, <code>species</code>, <code>epistasisMat</code> (epistasis matrix), <code>pleiotropyMat</code> (pleiotropy matrix), <code>q</code> (gene expression array), <code>biotic_phenotype</code>, <code>abiotic_phenotype</code>, <code>age</code>, <code>sex</code>, <code>energy</code>, <code>interaction_history</code>, and <code>W</code> (fitness).</p><p>The model object has the following fields that can be accessed with the <code>.</code> syntax and describe properties of species or the model: <code>ngenes</code>, <code>nphenotypes</code>, <code>growthrates</code>, <code>selectionCoeffs</code>, <code>ploidy</code>, <code>optvals</code> (optimal values), <code>optinds</code> (optval indices per generation), <code>mutProbs</code> (mutation probabilities), <code>mutMagnitudes</code> (mutation magnitudes), <code>N</code>, <code>E</code> (environmental noise), <code>generations</code>, <code>nspecies</code>, <code>migration_traits</code>, <code>vision_radius</code>, <code>check_fraction</code>, <code>migration_thresholds</code>, <code>step</code>, <code>biotic_phenotypes</code> (indices of biotic phenotypes per species), <code>abiotic_phenotypes</code>, <code>max_ages</code>, <code>food_sources</code>, <code>interactions</code>, <code>resources</code>, <code>recombination</code>, <code>initial_energy</code>.</p><p>You can collect data from agents and/or from the model object. To collect data from agents, use the <code>adata</code> keyword argument in the <code>runmodel</code> function, and to collect data from the model, use the <code>mdata</code> keyword. A complete description of the values these keywords take are at <a href="https://juliadynamics.github.io/Agents.jl/stable/tutorial/#.-Collecting-data">data collection section of the Agents.jl package</a>.</p><p>For example, we use the function below to count the number of individual per species:</p><pre><code class="language-jl hljs">&quot;Returns the size of each species.&quot;
function species_N(model::ABM)
  allagents = model.agents
  if length(allagents) == 0
    return fill(0, model.nspecies)
  else
    counts = countmap([a.species for a in values(model.agents)])
    output = fill(0, model.nspecies)
    for (k, v) in counts
      output[k] = v
    end
    return output
  end
end

using EvoDynamics

agentdata, modeldata, model = runmodel(&quot;parameters.yml&quot;, mdata=[species_N])</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorial/">« Tutorial</a><a class="docs-footer-nextpage" href="../example1/">Simple Wright-Fisher »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.5 on <span class="colophon-date" title="Tuesday 24 August 2021 21:11">Tuesday 24 August 2021</span>. Using Julia version 1.6.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
