<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · EvoDynamics.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">EvoDynamics.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li><a class="tocitem" href="../model_description/">Model description</a></li><li class="is-active"><a class="tocitem" href>Tutorial</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Basic-usage"><span>Basic usage</span></a></li><li><a class="tocitem" href="#Creating-simulation-parameter-files"><span>Creating simulation parameter files</span></a></li><li><a class="tocitem" href="#Collecting-data"><span>Collecting data</span></a></li><li><a class="tocitem" href="#Running-simulations-in-parallel"><span>Running simulations in parallel</span></a></li><li><a class="tocitem" href="#Modifying-the-sequence-of-events-and-specific-functions"><span>Modifying the sequence of events and specific functions</span></a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../example1/">Simple Wright-Fisher</a></li><li><a class="tocitem" href="../example2/">Predator prey</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/kavir1698/EvoDynamics.jl/blob/master/docs/src/tutorial.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial"><a class="docs-heading-anchor" href="#Tutorial">Tutorial</a><a id="Tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial" title="Permalink"></a></h1><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>Install using the following command in a Julia REPL.</p><pre><code class="language-julia hljs">]add EvoDynamics</code></pre><h2 id="Basic-usage"><a class="docs-heading-anchor" href="#Basic-usage">Basic usage</a><a id="Basic-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-usage" title="Permalink"></a></h2><p>Parameters of a model should be put in a julia file (.jl format) with the structure below.</p><p>See <a href="../example1/#Simple-Wright-Fisher">Simple Wright-Fisher</a> and <a href="../example2/#Predator-prey">Predator prey</a> for complete examples of parameter files.</p><pre><code class="language-jl hljs">## 1. Functions
...

## 2. Species parameters. A dictionary for each species

## 3. Model parameters as a dictionary.
</code></pre><p>The order of these sections is important because each section uses objects from its preceding sections. </p><p>Functions are used to create parameters that may change temporally and spatially. The following parameters are functions: bottleneck function which kills certain agents at certain times and spaces, optimal phenotype values which returns the optimal phenotype for a species at a given time and space, and environmental resources that may change over time.</p><p>You may create as many species as you want. Parameters of each species is a dictionary.</p><p>Model parameters is one dictionary that stores general parameters of the model, such as number of generations, space size, and species interaction parameters.</p><p>First, define your model parameters (here, we call it <code>parameters.jl</code>). <a href="../example1/#Simple-Wright-Fisher">Simple Wright-Fisher</a> and <a href="../example2/#Predator-prey">Predator prey</a> have examples of initiation parameters. See <a href="../model_description/#Model-description">Model description</a> for a description of each parameter.</p><p>We can the use the <code>runmodel</code> function to create a model from these parameters and run the simulation.</p><article class="docstring"><header><a class="docstring-binding" id="EvoDynamics.runmodel" href="#EvoDynamics.runmodel"><code>EvoDynamics.runmodel</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">runmodel(param_file::AbstractString; kwargs)</code></pre><p>Creates and runs a model given <code>parameters</code>. Returns a <code>DataFrame</code> of collected data, which are specified by <code>kwargs</code>.</p><p><strong>Keywords</strong></p><ul><li>adata=[] agent data to be collected. Either agent fields or functions that accept an agent as input can be put in the array. To aggregate collected data, provide tuples inside the array. For example, to collect mean and median fitness of individuals which is in field <code>W</code>, your array will be [(:W,mean), (:W,median)].</li><li>mdata=[mean<em>fitness</em>per<em>species, species</em>N] model data to be collected. By default, collects mean population fitness per species. Each row of the output DataFrame corresponds to all agents and each column is the value function applied to a field. The functions in the array are applied to the model object.</li><li>when=nothing The generations from which data are collected. By default collect at all steps.</li><li>replicates::Int = 0 Number of replicates per simulation.</li><li>parallel::Bool = false Whether to run replicates in parallel. If <code>true</code>, you should add processors to your julia session (e.g. by <code>addprocs(n)</code>) and define your parameters and <code>EvoDynamics</code> on all workers. To do that, add <code>@everywhere</code> before them. For example, <code>@everywhere EvoDynamics</code>.</li><li>seeds = optionally, provide an array of integers as seeds for each replicate.</li><li>agentstep=EvoDynamics.agent_step! Define your own agent stepping if you wish to change the sequence of events or change any one event.</li><li>modelstep=EvoDynamics.model_step!  </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/kavir1698/EvoDynamics.jl/blob/171005dbfa8ede113ad8bfc58770f369f992a17c/src/api.jl#L35-L50">source</a></section></article><pre><code class="language-jl hljs">using EvoDynamics
agentdata, modeldata, model = runmodel(&quot;parameters.jl&quot;)</code></pre><p>Within <code>runmodel</code>, before the parameters are used for constructing and ABM object, they are checked for correctness in type and shape using the <code>load_parameters</code> function.</p><article class="docstring"><header><a class="docstring-binding" id="EvoDynamics.load_parameters" href="#EvoDynamics.load_parameters"><code>EvoDynamics.load_parameters</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">load_parameters(paramfile::String)</code></pre><p>Reads the parameters from a parameter file (<code>.jl</code>), checks them for correct format, and returns the <code>model_parameters</code> dictionary object. See the online docs for the list of parameters and their correct values.</p><p>The output of this function can be loaded to the <code>model_initiation</code> function to create an ABM object. A user would normally not need to do these steps directly, because they are handled by the <code>runmodel</code> function. These will be useful for debugging and running the model in a different way.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/kavir1698/EvoDynamics.jl/blob/171005dbfa8ede113ad8bfc58770f369f992a17c/src/load_params.jl#L1-L7">source</a></section></article><p>And then, using the <code>model_initiation</code> function, an agent-based model (ABM) is constructed.</p><article class="docstring"><header><a class="docstring-binding" id="EvoDynamics.model_initiation" href="#EvoDynamics.model_initiation"><code>EvoDynamics.model_initiation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">model_initiation(dd)</code></pre><p>Innitializes the model and returns an ABM object that includes all the initial agents. The input of the function is a dictionary returned by the <code>load_parameters</code> function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/kavir1698/EvoDynamics.jl/blob/171005dbfa8ede113ad8bfc58770f369f992a17c/src/simulation.jl#L63-L68">source</a></section></article><p>Having and ABM object and parameters that define the run conditions, the <code>runmodel</code> function uses <code>run!</code> or <code>ensemblerun!</code> functions from the <code>Agents.jl</code> package to run the model and collect data.</p><h2 id="Creating-simulation-parameter-files"><a class="docs-heading-anchor" href="#Creating-simulation-parameter-files">Creating simulation parameter files</a><a id="Creating-simulation-parameter-files-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-simulation-parameter-files" title="Permalink"></a></h2><p>EvoDynamics.jl reads simulation parameters (<a href="../model_description/#Model-description">Model description</a>) from a julia file containing dictionaries and functions. This file can be populated manually using a text editor or from within a Julia session.</p><h2 id="Collecting-data"><a class="docs-heading-anchor" href="#Collecting-data">Collecting data</a><a id="Collecting-data-1"></a><a class="docs-heading-anchor-permalink" href="#Collecting-data" title="Permalink"></a></h2><p>The interface to the model is from the <code>runmodel</code> function (see <a href="#Tutorial">Tutorial</a>).</p><p>EvoDynamics.jl uses <a href="https://github.com/JuliaDynamics/Agents.jl">Agents.jl</a> underneath. See <a href="https://juliadynamics.github.io/Agents.jl/dev/">Agents.jl&#39;s documentation</a> for details about writing functions to collect any data during simulations. Here, we explain the specific implementation of the model.</p><p>There are two main objects from which you can collect data: and agent object of type <code>AbstractAgent</code> and a model object of type <code>ABM</code>. Both of these types are defined the <code>Agents.jl</code> package.</p><p>Agent object has the following fields that are individual specific: <code>id</code>, <code>pos</code>, <code>species</code>, <code>epistasisMat</code> (epistasis matrix), <code>pleiotropyMat</code> (pleiotropy matrix), <code>q</code> (gene expression array), <code>biotic_phenotype</code>, <code>abiotic_phenotype</code>, <code>age</code>, <code>sex</code>, <code>energy</code>, <code>interaction_history</code>, and <code>W</code> (fitness).</p><p>The model object has the following fields that can be accessed with the <code>.</code> syntax and describe properties of species or the model: <code>ngenes</code>, <code>nphenotypes</code>, <code>growthrates</code>, <code>selectionCoeffs</code>, <code>ploidy</code>, <code>optvals</code> (optimal values), <code>optinds</code> (optval indices per generation), <code>mutProbs</code> (mutation probabilities), <code>mutMagnitudes</code> (mutation magnitudes), <code>N</code>, <code>E</code> (environmental noise), <code>generations</code>, <code>nspecies</code>, <code>migration_traits</code>, <code>vision_radius</code>, <code>check_fraction</code>, <code>migration_thresholds</code>, <code>step</code>, <code>biotic_phenotypes</code> (indices of biotic phenotypes per species), <code>abiotic_phenotypes</code>, <code>max_ages</code>, <code>food_sources</code>, <code>interactions</code>, <code>resources</code>, <code>recombination</code>, <code>initial_energy</code>.</p><p>You can collect data from agents and/or from the model object. To collect data from agents, use the <code>adata</code> keyword argument in the <code>runmodel</code> function, and to collect data from the model, use the <code>mdata</code> keyword. A complete description of the values these keywords take are at <a href="https://juliadynamics.github.io/Agents.jl/stable/tutorial/#.-Collecting-data">data collection section of the Agents.jl package</a>.</p><p>For example, we use the function below to count the number of individual per species:</p><pre><code class="language-jl hljs">&quot;Returns the size of each species.&quot;
function species_N(model::ABM)
  allagents = model.agents
  if length(allagents) == 0
    return fill(0, model.nspecies)
  else
    counts = countmap([a.species for a in values(model.agents)])
    output = fill(0, model.nspecies)
    for (k, v) in counts
      output[k] = v
    end
    return output
  end
end

using EvoDynamics

agentdata, modeldata, model = runmodel(&quot;parameters.yml&quot;, mdata=[species_N])</code></pre><h2 id="Running-simulations-in-parallel"><a class="docs-heading-anchor" href="#Running-simulations-in-parallel">Running simulations in parallel</a><a id="Running-simulations-in-parallel-1"></a><a class="docs-heading-anchor-permalink" href="#Running-simulations-in-parallel" title="Permalink"></a></h2><p>You can run replicate simulation in parallel. To that end, you need to add processors, and import EvoDynamics.jl and your parameters files on all cores:</p><pre><code class="language-julia hljs">using Distributed
addprocs(4)
@everywhere using EvoDynamics
@everywhere param_file = &quot;params.jl&quot;
@everywhere EvoDynamics.load_parameters(param_file)
adata, mdata, models = runmodel(param_file, replicates=10, parallel=true)</code></pre><h2 id="Modifying-the-sequence-of-events-and-specific-functions"><a class="docs-heading-anchor" href="#Modifying-the-sequence-of-events-and-specific-functions">Modifying the sequence of events and specific functions</a><a id="Modifying-the-sequence-of-events-and-specific-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Modifying-the-sequence-of-events-and-specific-functions" title="Permalink"></a></h2><p>If you want to change the sequence of events within each time-step, you can do so by providing your own stepping function to the <code>runmodel</code> function. With this flexibility, you can also keep the sequence as it is, but change one specific event within the sequence by using a different function for it. A good staring point would be to copy the <code>agent_step!</code> function from the package (at <code>src/simulation.jl</code>).</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../model_description/">« Model description</a><a class="docs-footer-nextpage" href="../example1/">Simple Wright-Fisher »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Wednesday 16 November 2022 11:02">Wednesday 16 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
