<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Model description · EvoDynamics.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">EvoDynamics.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li class="is-active"><a class="tocitem" href>Model description</a><ul class="internal"><li><a class="tocitem" href="#Parameters"><span>Parameters</span></a></li><li><a class="tocitem" href="#Simulation-outline"><span>Simulation outline</span></a></li></ul></li><li><a class="tocitem" href="../tutorial/">Tutorial</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../example1/">Simple Wright-Fisher</a></li><li><a class="tocitem" href="../example2/">Predator prey</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Model description</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Model description</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/kavir1698/EvoDynamics.jl/blob/master/docs/src/model_description.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Model-description"><a class="docs-heading-anchor" href="#Model-description">Model description</a><a id="Model-description-1"></a><a class="docs-heading-anchor-permalink" href="#Model-description" title="Permalink"></a></h1><h2 id="Parameters"><a class="docs-heading-anchor" href="#Parameters">Parameters</a><a id="Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Parameters" title="Permalink"></a></h2><p>All parameters of a model are written in a julia (<code>.jl</code>) file. Parameters are divided into two sections: species specific and model specific. Parameters for each species and for the model are written in separate dictionary (<code>Dict</code>) objects. The keys of the dictionary should be Symbols (use colon : before the names).</p><p>An easy way to parameterize a model is to copy an existing set of parameters from the Examples section and modify them.</p><h3 id=".-Species-specific-parameters"><a class="docs-heading-anchor" href="#.-Species-specific-parameters">1. Species specific parameters</a><a id=".-Species-specific-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#.-Species-specific-parameters" title="Permalink"></a></h3><p>Each species should have the following parameters in a dictionary objection. The order that you write these parameters does not matter.</p><ul><li><strong>name</strong>: A <code>String</code> as the name of the species.</li><li><strong>number_of_genes</strong>: An <em>integer</em> for number of genes that the species has.</li><li><strong>ploidy</strong>: Either 1 for haploid or 2 for diploid genomes. Diploids may undergo recombination.</li><li><strong>number_of_phenotypes</strong>: An <em>integer</em> for the number of phenotypes that the species has.</li><li><strong>abiotic_phenotypes</strong>: An <em>array of integers</em> (e.g. &quot;[1,2]&quot;) specifying abiotic phenotypes among all phenotypes. Abiotic phenotypes determine how the species interacts with the environment.</li><li><strong>biotic_phenotypes</strong>: An <em>array of integers</em> (e.g. &quot;[3]&quot;) specifying biotic phenotypes among all phenotypes. Biotic phenotypes determine how the species interacts with other individuals from the same or different species.</li><li><strong>migration_phenotype</strong>: An <em>integer</em> specifying the phenotype that corresponds to the migration trait. If the species does not migrate, put 0.</li><li><strong>migration_threshold</strong>: The phenotypic value of the migration phenotype after which an individual can migrate.</li><li><strong>vision_radius</strong>: A <em>number</em> determining the radius of neighboring sites that the agent can see before migration.</li><li><strong>check_fraction</strong>: A <em>floating number</em> between 0 and 1 showing the fraction of the visible sites to the agent that it can check and decide whether to migrate to.</li><li><strong>epistasis_matrix</strong>: An epistasis matrix is of size <span>$l \times l$</span>, where <em>l</em> is the product of <em>number of genes</em> and <em>ploidy</em>. Epistasis matrix specifies the direction (positive or negative) and size of effect of one locus on other loci. For example, if at row 1 and column 2 is a value 0.2, it means that locus 1 affects locus 2 by increasing the effect of locus 2 (because its positive) with 20% of the effect of locus 1.</li><li><strong>pleiotropy_matrix</strong>: A <em>binary matrix</em> (0s and 1s) with size <em>number of phenotypes</em> times <em>l</em>. The pleiotropy matrix specifies the phenotypes that each locus affects.</li><li><strong>expression_array</strong>: A <em>vector</em> of size <em>l</em> that represent the expression amount of each locus determining its effect size.</li><li><strong>growth_rate</strong>: Mean of a Poisson distribution for number of offsprings per reproduction. This number is fully realized when fitness of a haploid individual is 1, or the distance between the biotic phenotypes of two diploid individuals is 0. Otherwise, actual growth rate is a fraction of this value. </li><li><strong>selection_coefficient</strong>: A number between 0 and 1 that determines the importance of fitness. 0 is be a model without selection. If you want time-variable selection coefficient, provide a vector as long as number of generations plus 1 (to account for generation zero).</li><li><strong>phenotype_contribution_to_fitness</strong>: The relative contribution of each <strong>abiotic phenotype</strong> to fitness. This parameter can be <code>nothing</code> to denote equal contribution of all abiotic phenotypes, or it can be a vector of numbers with as many elements as there are abiotic phenotypes.</li><li><strong>mutation_probabilities</strong>: A <em>vector of three numbers</em> each of which specifies the probability for a different type of mutations: mutation probability of the <em>expression\</em>array<em>, _pleiotropy\</em>matrix<em>, and _epistasis\</em>matrix_, respectively.</li><li><strong>mutation_magnitudes</strong>: A <em>vector of numbers</em> with the same size as <em>mutation\</em>probabilities_ that determines the magnitude of mutation for each of the three categories. Specifically, the numbers are the variances of normal distributions with mean 0 for expression array and epistasis matrices, and probability of changing a 0 and 1 in in the pleiotropy matrix. When mutating an offspring, it first checks whether there will be a mutation (<em>mutation\</em>probabilities<em>), and if positive, a mutation with a magnitude determined by _mutation\</em>magnitudes_ occurs.</li><li><strong>N</strong>: A <em>vector of integers</em> for the initial number of individuals at each site.</li><li><strong>environmental_noise</strong>: A number for the variance of a normal distribution with mean 0 that will be added to the phenotypes.</li><li><strong>optimal_phenotypes</strong>: A <em>vector of matrices</em>. The vector holds optimal phenotypes at each time step. Each matrix, is the optimal phenotypes for each site in space. The optimal phenotype at each site is a vector as long as <code>abiotic_phenotypes</code>. The vector is as long as model steps (generations) plus 1 (for time zero).</li><li><strong>age</strong>: An <em>integer</em> for maximum age of individuals of this species.</li><li><strong>reproduction_start_age</strong>: The age at which individuals can reproduce.</li><li><strong>reproduction_end_age</strong>: The age after which individuals cannot reproduce.</li><li><strong>recombination</strong>: Mean of a Poisson distributions for number of crossing overs per sexual reproduction.</li><li><strong>initial_energy</strong>: A parameter for parental care of infants. Values more than 0 indicate that newly born individuals can survive for a number of times without requiring food from the environment/other species. The consumption rate is always constant for all species at one unit per time step. Use with care. Having initial energy larger than zero can lead to infinite population growth because agents without food can reproduce and their offsprings also reproduce without food. For example, this happens when start age of reproduction is 1 and initial energy is larger than 0.</li><li><strong>bottlenecks</strong>: An vector of matrices each of which has the same size as space and contains the probability of death at each site. The vector is as long as model steps (generations) plus 1 (for time zero). Use this if you want to impose a killing event in the model at specific times and places. Otherwise, just use 0.0 for all the values in matrices.</li><li><strong>mating_scheme</strong>: One of the following: -1, 0, 1. Only used in sexual reproduction. Zero means randomly paired couple are likely to have as many offsprings as any other pair. Their phenotypes does not affect their reproduction success. -1 is disassortative mating, where the more different the phenotypes of the pair, the more children they have. 1 is the opposite, assortative mating.</li><li><strong>abiotic_variance</strong>: Variance of a normal distribution used in determining the phenotypic distance of agents to the optimal environmental phenotypes. The larger the variance, the less important is the distance.</li><li><strong>biotic_variance</strong>: Same as <em>abiotic\</em>variance_ but for determining the biotic phenotypic distance between two individuals (used in any kind of interaction). Larger values mean that all pairs are equally likely to interact, regardless of their phenotype difference.</li></ul><h3 id=".-Model-parameters"><a class="docs-heading-anchor" href="#.-Model-parameters">2. Model parameters</a><a id=".-Model-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#.-Model-parameters" title="Permalink"></a></h3><p>This dictionary should be named &quot;model_parameters&quot;.</p><ul><li><strong>species</strong>: A vector containing the name of the species <em>dictionaries</em>.</li><li><strong>generations</strong>: An <em>integer</em> for the number of steps the model will run.</li><li><strong>space</strong>: A <em>tuple of two integers</em> (e.g. <code>(3,2)</code>) that determines the size space. If you do not want any spatial structure, use <code>(1,1)</code>.</li><li><strong>metric</strong>: Either &quot;chebyshev&quot; or &quot;euclidian&quot;. Determines how many neighbors a space site has. &quot;chebyshev&quot; metric means that the r-neighborhood of a position are all positions within the hypercube having side length of 2*floor(r) and being centered in the origin position. &quot;euclidean&quot; metric means that the r-neighborhood of a position are all positions whose cartesian indices have Euclidean distance ≤ r from the cartesian index of the given position.</li><li><strong>periodic</strong>: _Boolean__ (true or false) to determine whether boundaries of the space are connected or not.</li><li><strong>resources</strong>: A <em>vector of matrices</em> where each matrix is the size of space and determines the amount of resources per site. The vector is as long as model steps (generations) plus 1 (for time zero).</li><li><strong>interactions</strong>: A species-species interaction <em>matrix of floating numbers</em> determining how individuals from different species interact. Each value  is strength of interaction (between 0 and 1). Sign (+/-) is the direction of interaction where positive means similar individuals interact more strongly and negative is dissimilar ones tend to interact more.</li><li><strong>food_sources</strong>: A species-species food <em>matrix of floating numbers</em> determining what each species feeds on (consumption rate). Non-zero diagonal means the food resource is from the environment. Off-diagonals mean an species (in the rows) feeds on another species (in the columns). Numbers can be zero or any positive number. The magnitude of the number determines how much energy the predator gains by eating the prey. All species use 1 unit of energy per time step.</li><li><strong>seed</strong>: Either an <em>integer</em> or <code>nothing</code> for a random number.</li></ul><h2 id="Simulation-outline"><a class="docs-heading-anchor" href="#Simulation-outline">Simulation outline</a><a id="Simulation-outline-1"></a><a class="docs-heading-anchor-permalink" href="#Simulation-outline" title="Permalink"></a></h2><p>The simulations are fully agent-based, meaning that agents do not receive any model-level knowledge for what happens to them. The following steps happen in order to agents that are activated in a random order.</p><ol><li>Grow one step older.</li><li>Migrate.</li><li>Burn energy.</li><li>Eat from the environment if the species can.</li><li>Interact with other individuals.</li><li>Reproduce.</li><li>Survive. Agents die if they are too old, or do not have enough energy, or by chance given its fitness.</li><li>Kill agents will a probability given in <code>bottlenecks</code>.</li></ol><h3 id="Migration"><a class="docs-heading-anchor" href="#Migration">Migration</a><a id="Migration-1"></a><a class="docs-heading-anchor-permalink" href="#Migration" title="Permalink"></a></h3><p>If a species have the capability to migrate (i.e. <em>migration phenotype</em> is not 0), then at each time step, it is checked whether the phenotypic value of the migration phenotype is above <em>migration threshold</em>. If it is, then the agent checks a random <em>check fraction</em> of the sites neighboring it and moves to the most suitable site. Suitability is calculated by comparing the agent&#39;s abiotic phenotype and the optimal abiotic phenotypic value for each site. If the checked neighbors have worse conditions than the current site, the agent does not move.</p><h3 id="Burn-energy"><a class="docs-heading-anchor" href="#Burn-energy">Burn energy</a><a id="Burn-energy-1"></a><a class="docs-heading-anchor-permalink" href="#Burn-energy" title="Permalink"></a></h3><p>Agent&#39;s energy is reduced by one unit.</p><h3 id="Eat"><a class="docs-heading-anchor" href="#Eat">Eat</a><a id="Eat-1"></a><a class="docs-heading-anchor-permalink" href="#Eat" title="Permalink"></a></h3><p>If the agent is able to eat from the environment (the diagonal of <em>food sources</em> at the corresponding row and column is non-zero) and there is any environmental resources left at the agent&#39;s site, then its energy level is boosted by as much as the value in the corresponding element at <em>food sources</em>.</p><h3 id="Interacting-with-other-individuals"><a class="docs-heading-anchor" href="#Interacting-with-other-individuals">Interacting with other individuals</a><a id="Interacting-with-other-individuals-1"></a><a class="docs-heading-anchor-permalink" href="#Interacting-with-other-individuals" title="Permalink"></a></h3><p>At each time, the agent interacts with as many agents as there are species in the model. If there are fewer agents in the site than number of species, the agent interacts with all of them. Otherwise, it interacts with at most one randomly picked individual from each species. If most of the individuals at the site are from the first species, then it is more likely that the agent interacts with an individual from the first species and with no individual from the second species. This set up allows interactions between species be dependent on the population size of the species at each site.</p><p>For each pair of interacting individuals, we first check whether one individual feeds on the other one. If so, the hunt is successful with a probability proportional to the average phenotypic distance between the biotic phenotypes of the two individuals.</p><p>Phenotypic distance between two species depends on the sign of the corresponding element in <em>interactions</em> matrix. If it is negative, then the hunt is more successful if the average phenotype of the two individuals is more different.</p><p>The phenotypic distance between two individuals is the average distance between all pairs of biotic phenotypes between the two individuals. To calculate the distance between two phenotypic values, we use the following formula:</p><p>$ | 0.5 - \text{cdf}(\text{Normal}(\text{ph1},  1), \text{ph2})| $</p><p>, where cdf is cumulative density function, Normal is a normal distribution with mean ph1 (phenotypic value 1) and variance 1, and ph2 is phenotypic value 2.</p><p>If the two individuals are not predator-prey, and they are from the same species but different sexes and they are both in reproduction age, then they are marked to reproduce in the next stage. Otherwise, if they are from different species and the two species interact with each other (the corresponding values in the <em>interactions</em> matrix is non-zero) they interact with each other. If the individual 1 to 2 has a positive value in the <em>interactions</em> matrix, then it increases the fitness of individual 2. If the element is zero, it does not increase or decrease the fitness of individual 2. And it the element is negative, it decreases the fitness of individual 2. Similarly, individual 2 can affect the fitness of individual 1. The two interactions do not need to be symmetric. The magnitude of change in the fitness due to interaction is determined by the phenotypic distance between the two individuals and interaction value in the <em>interactions</em> matrix.</p><h3 id="Reproduction"><a class="docs-heading-anchor" href="#Reproduction">Reproduction</a><a id="Reproduction-1"></a><a class="docs-heading-anchor-permalink" href="#Reproduction" title="Permalink"></a></h3><p>If an individual is haploid and is in reproduction age, it reproduces an identical offspring to itself except that the expression array, pleiotropy matrix, and the epistasis matrix of the offspring will mutate given the probabilities and magnitudes in the <em>mutation probabilities</em> and <em>mutation magnitudes</em> matrices. The number of offsprings is a random number from a Poisson distribution with a mean equal to the species&#39; growth rate times the fitness of the individual.</p><p>If two diploid individuals are mated to reproduce, their reproduction success is proportional to their phenotypic distance (depending on <code>mating\_scheme</code>). Number of offsprings is a random number from a Poisson distribution with mean equal to the reproduction success of the two individuals times the growth rate of the species.</p><p>Each offspring of diploid individuals inherits a gamete from each parent. A gamete is a half of the expression array, pleiotropy matrix, and epistasis matrix. If recombination is allowed (&gt; 0), then the gametes undergo crossing over. The number of crossing overs is a random number from a Poisson distribution with mean equal to the <em>recombination</em> parameter.</p><h4 id="Survival"><a class="docs-heading-anchor" href="#Survival">Survival</a><a id="Survival-1"></a><a class="docs-heading-anchor-permalink" href="#Survival" title="Permalink"></a></h4><p>At each step, the agent may die due to several factors. First, it dies if it has negative energy (has not been able to eat enough). Second, it dies if it is too old (&gt; max age). Third, it dies with a probability negatively correlated to its fitness. This probability is adjusted to the <em>selection coefficient</em>. If the selection coefficient is zero, then the individual does not die. If it is one, the fitness determines survival 100%.</p><h4 id="Fitness"><a class="docs-heading-anchor" href="#Fitness">Fitness</a><a id="Fitness-1"></a><a class="docs-heading-anchor-permalink" href="#Fitness" title="Permalink"></a></h4><p>Fitness of individuals is their abiotic fitness, i.e. their abiotic phenotypic distance to the optimal phenotypes in the environment. As such, their fitness can be different at different sites.</p><h4 id="Mutation"><a class="docs-heading-anchor" href="#Mutation">Mutation</a><a id="Mutation-1"></a><a class="docs-heading-anchor-permalink" href="#Mutation" title="Permalink"></a></h4><p>Mutation can happen at three levels: changing the expression of each gene, changing the pleiotropy matrix, and changing the epistasis interactions between genes. The probability that a mutation occurs at each of these levels is controlled by parameter <em>mutation probabilities</em>. And size of mutations when they occur are controlled by the <em>mutation magnitudes</em> parameters, which are the variances of a normal distribution. A mutation in gene expression and epistasis matrix is carried by adding a random number from a normal distribution to the existing values. The pleiotropy matrix is mutated by switching 0s and 1s with the probability given in the third element of <em>mutation magnitudes</em>.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Introduction</a><a class="docs-footer-nextpage" href="../tutorial/">Tutorial »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 8 November 2022 09:47">Tuesday 8 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
